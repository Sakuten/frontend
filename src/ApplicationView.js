import React, { Component } from 'react';
import { inject, observer } from 'mobx-react';

@inject('event')
@observer
export class ApplicationView extends Component {
  render() {
    return (
      <div>
        <h1>Logged in as {this.props.user.get('username')}</h1>
        <button onClick={this.props.event.credential.logout}>Logout</button>
      </div>
    );
  }
}
