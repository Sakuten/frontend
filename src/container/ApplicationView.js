import React from 'react'
import styled from 'styled-components'
import { inject, observer } from 'mobx-react'

import ApplicationList from '../component/ApplicationList'
import GroupMemberList from '../component/GroupMemberList'
import GroupMemberButton from '../component/GroupMemberButton'
import ClassroomSelect from '../component/ClassroomSelect'
import LotterySelect from '../component/LotterySelect'
import Button from '../component/Button'

import cloud from '../cloud.svg'

const ButtonWrap = styled.button`
  margin: 20px 0;
  text-align: center;
  line-height: 2rem;
  font-size: 2rem;
  border: none;
  width: 100%;
  background-color: transparent;

  position: relative;
`

const Selection = styled.div`
  background-color: ${props => props.theme.header_color_lower};
  width: 80vw;
  height: 70%;
  padding: 30px;
`

const Title = styled.h3`
  font-size: 1.5rem;
  color: #000;
`

const Indent = styled.div`
  margin: 10px;
`

const Container = styled.div`
  margin-top: 20px;
`

const Cloud = styled.div`
  position: absolute;
  width: 60vw;
  right: -20%;
  z-index: 20;
`

const ApplicationView = ({user, application, event}) => {
  const {
    onQRError
  } = event.credential

  const {
    onChangeClassroom,
    onChangeLottery,
    onApply,
    onCancel,
    onAddGroupMember,
    onRemoveGroupMember
  } = event.application

  return (
    <div data-test='applicationview'>
      <Container>
        <Cloud>
          <img src={cloud} />
        </Cloud>
        <Selection>
          <Title>クラス選択</Title>
          <Indent>
            <ClassroomSelect list={application.classroomList} value={application.classroom} onChange={onChangeClassroom} />
          </Indent>
          <Title>応募する公演</Title>
          <Indent>
            <LotterySelect classroom={application.classroom} list={application.lotteryList} value={application.lottery} onChange={onChangeLottery} />
          </Indent>
          <Title>一緒に応募する</Title>
          <Indent>
            <GroupMemberList list={application.groupMemberList} onRemove={onRemoveGroupMember} />
            <GroupMemberButton onAdd={onAddGroupMember} onError={onQRError}>他の人を追加</GroupMemberButton>
          </Indent>
        </Selection>
        <ButtonWrap onClick={onApply}>
          <Button>
            申し込む
          </Button>
        </ButtonWrap>
        <Selection>
          <Title>応募一覧</Title>
          <Indent>
            <ApplicationList list={application.applicationList} onCancel={onCancel} />
          </Indent>
        </Selection>
      </Container>
    </div>
  )
}

export default inject('event')(observer(ApplicationView))
