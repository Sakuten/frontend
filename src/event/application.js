import {applyLottery, cancelLottery} from '../api/operation'

export class ApplicationObject {
  constructor (store) {
    this.store = store
    this.onUpdate()
  }

  onUpdate = async () => {
    await Promise.all([
      this.store.application.fetchClassroomList(),
      this.store.application.fetchLotteryList()
    ])
  }

  onChangeClassroom = (classroom) => {
    this.store.application.setClassroom(classroom)
  }

  onChangeLottery = (lottery) => {
    this.store.application.setLottery(lottery)
  }

  onAddGroupMember = (secretId) => {
    this.store.application.addGroupMember(secretId)
  }

  onRemoveGroupMember = (idx) => {
    this.store.application.removeGroupMemberWithIdx(idx)
  }

  onApply = async () => {
    await applyLottery(this.store.application.lottery, this.store.application.groupMemberList, this.store.credential.token)
    await this.store.fetchStatus()
  }

  onCancel = async (id) => {
    await cancelLottery(id, this.store.credential.token)
    await this.store.fetchStatus()
  }
}
