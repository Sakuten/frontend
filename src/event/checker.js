import {getPublicId, checkSecretIdStatus} from '../api/operation'
import {extractId} from '../util/extractId'

export class CheckerObject {
  constructor (store) {
    this.store = store
    this.lastScan = ''
  }

  onChangeClassroom = (classroom) => {
    this.store.checker.setClassroom(classroom)
  }

  onQRScan = async (scanUri) => {
    if (scanUri) {
      if (this.lastScan === scanUri) {
        return
      }
      this.lastScan = scanUri
      const secretId = extractId(scanUri)
      if (!secretId) {
        this.store.error.addError(101, 'Invalid QR Code')
        return
      }

      this.store.checker.setLastStatus('取得中')
      this.store.checker.setPublicId('')

      const status = await this.store.error.ignoring([19, 6], (callback) => {
        return checkSecretIdStatus(this.store.checker.classroom, secretId, this.store.credential.token)
          .then(resp => {
            return resp.data['status']
          })
          .catch(e => {
            const codeStatuses = {
              19: '応募していません',
              6: '時間外です'
            }
            return codeStatuses[e.response.data.code] || JSON.stringify(e.response.data)
          })
          .then(callback)
      })

      const resp = await getPublicId(secretId, this.store.credential.token)

      this.store.checker.setLastStatus(status)
      this.store.checker.setPublicId(resp.data['public_id'])
    }
  }
}
