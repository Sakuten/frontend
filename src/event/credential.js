import {fetchApi} from '../api'

export class CredentialObject {
  constructor(store) {
    this.store = store;
  }

  login = () => {
    fetchApi('auth/', {
      method: 'post',
      headers: {
        'Content-Type': 'application/json'
      },
      data: {
        password: this.store.credential.password,
        username: this.store.credential.username
      }
    }).then(async (response) => {
      const json = response.data
      if ('token' in json) {
        this.store.credential.setToken(json.token)
        this.store.credential.clearPassword()
        this.fetchStatus()
      } else { throw Error('Invalid response returned') }
    })
  }

  fetchStatus = () => {
    fetchApi(`api/status`, {
      method: 'get',
      headers: {
        'Authorization': 'Bearer ' + this.store.credential.token
      }
    })
      .then(response => {
        this.store.credential.setStatus(response.data.status)
      })
      .catch(error => {
        if (error.response && error.response.data.message === 'Unauthorized') {
          this.logout()
        } else {
          throw error
        }
      })
  }

  logout = () => {
    this.store.credential.setToken('')
  }

  onChangeUsername = (username) => {
    username = username.trim()
    this.store.credential.setUsername(username)
  }

  onChangePassword = (password) => {
    password = password.trim()
    this.store.credential.setPassword(password)
  }
}
