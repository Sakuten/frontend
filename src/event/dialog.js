export class DialogObject {
  constructor (store) {
    this.store = store
  }

  onClose = () => {
    this.store.dialog.close()
  }

  onOpen = (title, content, buttonText = 'OK') => {
    this.store.dialog.setTitle(title)
    this.store.dialog.setContent(content)
    this.store.dialog.setButtonText(buttonText)
    this.store.dialog.open()
  }
}
