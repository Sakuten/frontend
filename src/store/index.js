import {CredentialObject} from './credential'
import {ApplicationObject} from './application'
import { RouterStore } from 'mobx-react-router'
import {ErrorObject} from './error'

export class Store {
  constructor () {
    this.credential = new CredentialObject()
    this.application = new ApplicationObject()
    this.router = new RouterStore()
    this.error = new ErrorObject()
  }

  // Called in Event's constructor
  fetchStatus = async () => {
    await Promise.all([
      this.credential.fetchStatus(),
      this.application.fetchApplicationList(this.credential.token)
    ])
  }
}
