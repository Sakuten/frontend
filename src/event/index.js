import {CredentialObject} from './credential'
import {ApplicationObject} from './application'
import {CheckerObject} from './checker'
import {ErrorObject} from './error'
import {DialogObject} from './dialog'

import {applyLottery, cancelLottery} from '../api/operation'

export class Event {
  constructor (store) {
    this.store = store
    this.error = new ErrorObject(store)
    this.credential = new CredentialObject(store)
    this.application = new ApplicationObject(store)
    this.dialog = new DialogObject(store)
    this.checker = new CheckerObject(store)

    // This can cause network error, to be caught in ErrorObject
    // So fetchStatus() have to be placed in here
    if (this.store.credential.isLoggedIn) { this.store.fetchStatus() }
  }

  onApplyLottery = async () => {
    this.dialog.onOpen('応募しています', 'しばらくお待ちください', 'お待ちください', false)
    try {
      await applyLottery(this.store.application.lottery, this.store.application.groupMemberList.map(pair => pair[0]), this.store.credential.token)
    } catch (err) {
      this.dialog.onClose()
      throw err
    }
    this.dialog.onOpen('応募しました', '発表をお待ちください', 'OK')
    await this.store.fetchStatus()
    if (this.store.credential.isUsedByStaff) {
      this.store.credential.logout()
    }
  }

  onCancelApplication = async (id) => {
    this.dialog.onOpen('キャンセルしています', 'しばらくお待ちください', 'お待ちください', false)
    try {
      await cancelLottery(id, this.store.credential.token)
    } catch (err) {
      this.dialog.onClose()
      throw err
    }
    this.dialog.onOpen('キャンセルしました', '応募は取り消されました', 'OK')
    await this.store.fetchStatus()
    if (this.store.credential.isUsedByStaff) {
      this.store.credential.logout()
    }
  }
}
