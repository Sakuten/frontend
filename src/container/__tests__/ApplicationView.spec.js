import React from 'react'
import { shallow, mount } from 'enzyme'
import deepAssign from 'deep-assign'
import { Provider } from 'mobx-react'
import ApplicationView from '../ApplicationView'
import {Store} from '../../store'
import {Event} from '../../event'

const setup = (propOverrides, storeOverrides, eventOverrides, isShallow = true) => {
  const store = deepAssign(new Store(), storeOverrides)
  const event = deepAssign(new Event(store), eventOverrides)
  const props = Object.assign({
    credential: store.credential,
    application: store.application
  }, propOverrides)

  const wrapper = (isShallow ? shallow : mount)(<Provider event={event}><ApplicationView {...props} /></Provider>)

  return {
    props,
    wrapper,
    store,
    event,
    title: wrapper.find('[data-test="applicationview-title"]')
  }
}

describe('containers', () => {
  describe('ApplicationView', () => {
    it('render', () => {
      const { wrapper } = setup()
      expect(wrapper).toMatchSnapshot()
    })

    it('renders classroom options', async () => {
      const { event, wrapper } = setup({}, {}, {}, false)
      await event.application.onUpdate()
      expect(wrapper.render().find('[data-test="classroom-option"]').length).toBe(8)
    })

    it('renders application list', async () => {
      const { store, wrapper } = setup({}, {}, {}, false)
      await store.fetchStatus()
      expect(wrapper.render().find('[data-test="applicationlist-application"]').length).toBe(1)
    })
  })
})
