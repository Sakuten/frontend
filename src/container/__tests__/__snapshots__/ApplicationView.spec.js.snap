// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers ApplicationView render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    event={
      Event {
        "application": ApplicationObject {
          "onApply": [Function],
          "onCancel": [Function],
          "onChangeClassroom": [Function],
          "onChangeLottery": [Function],
          "onUpdate": [Function],
          "store": Store {
            "application": ApplicationObject {},
            "credential": CredentialObject {
              "recaptchaResponse": "",
              "secretId": "",
              "status": Object {},
              "token": "",
              Symbol(mobx did run lazy initializers): true,
              Symbol(mobx administration): ObservableObjectAdministration$$1 {
                "defaultEnhancer": [Function],
                "keysAtom": Atom$$1 {
                  "diffValue": 0,
                  "isBeingObserved": false,
                  "isPendingUnobservation": false,
                  "lastAccessedBy": 0,
                  "lowestObserverState": 2,
                  "name": "CredentialObject@1.keys",
                  "observers": Set {},
                },
                "name": "CredentialObject@1",
                "target": [Circular],
                "values": Map {
                  "secretId" => "",
                  "recaptchaResponse" => "",
                  "token" => "",
                  "status" => Map {},
                  "isLoggedIn" => false,
                  "isAbleToAuthenicate" => false,
                },
              },
            },
            "error": ErrorObject {},
            "fetchStatus": [Function],
            "router": RouterStore {
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "history": null,
              "push": [Function],
              "replace": [Function],
            },
          },
        },
        "credential": CredentialObject {
          "onChangeRecaptchaResponse": [Function],
          "onLogin": [Function],
          "onLogout": [Function],
          "onQRError": [Function],
          "onQRScan": [Function],
          "store": Store {
            "application": ApplicationObject {},
            "credential": CredentialObject {
              "recaptchaResponse": "",
              "secretId": "",
              "status": Object {},
              "token": "",
              Symbol(mobx did run lazy initializers): true,
              Symbol(mobx administration): ObservableObjectAdministration$$1 {
                "defaultEnhancer": [Function],
                "keysAtom": Atom$$1 {
                  "diffValue": 0,
                  "isBeingObserved": false,
                  "isPendingUnobservation": false,
                  "lastAccessedBy": 0,
                  "lowestObserverState": 2,
                  "name": "CredentialObject@1.keys",
                  "observers": Set {},
                },
                "name": "CredentialObject@1",
                "target": [Circular],
                "values": Map {
                  "secretId" => "",
                  "recaptchaResponse" => "",
                  "token" => "",
                  "status" => Map {},
                  "isLoggedIn" => false,
                  "isAbleToAuthenicate" => false,
                },
              },
            },
            "error": ErrorObject {},
            "fetchStatus": [Function],
            "router": RouterStore {
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "history": null,
              "push": [Function],
              "replace": [Function],
            },
          },
        },
        "store": Store {
          "application": ApplicationObject {},
          "credential": CredentialObject {
            "recaptchaResponse": "",
            "secretId": "",
            "status": Object {},
            "token": "",
            Symbol(mobx did run lazy initializers): true,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "CredentialObject@1.keys",
                "observers": Set {},
              },
              "name": "CredentialObject@1",
              "target": [Circular],
              "values": Map {
                "secretId" => "",
                "recaptchaResponse" => "",
                "token" => "",
                "status" => Map {},
                "isLoggedIn" => false,
                "isAbleToAuthenicate" => false,
              },
            },
          },
          "error": ErrorObject {},
          "fetchStatus": [Function],
          "router": RouterStore {
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "history": null,
            "push": [Function],
            "replace": [Function],
          },
        },
      }
    }
  >
    <inject-ApplicationView-with-event
      application={ApplicationObject {}}
      user={Object {}}
    />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "application": ApplicationObject {},
      "user": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "application": ApplicationObject {},
        "user": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
